### PR TITLE
Add preallocating constructor for BitSet(::AbstractVector)

### PR DESCRIPTION
**Pull request by LilithHafner**
Ideally, we would dispatch to the new method for any efficient iterable that can be iterated twice; the closest type I could find was `AbstractVector`.

Benchmarking code:
```julia
using BenchmarkTools

function f(n)
    for x in 0:n
        for y in 0:n
            v = rand(1:10^x, 10^y)
            t0 = @belapsed invoke(BitSet, Tuple{Any}, x) setup=(x=rand(1:10^$x, 10^$y))
            t1 = @belapsed invoke(BitSet, Tuple{AbstractVector}, x) setup=(x=rand(1:10^$x, 10^$y))
            println((x, y, t1/t0))
        end
    end
end

f(4)
```
Results:
```
(0, 0, 0.5087067955287577)
(0, 1, 0.5685897660058082)
(0, 2, 0.8094049703272072)
(0, 3, 0.9940489071629518)
(0, 4, 1.0221360750603967)
(1, 0, 0.5116265190362661)
(1, 1, 0.5650601757040765)
(1, 2, 0.8121002300412222)
(1, 3, 0.9947459646842162)
(1, 4, 1.0287911101133684)
(2, 0, 0.5057934999240832)
(2, 1, 0.5449234307575512)
(2, 2, 0.5740818095718663)
(2, 3, 0.6302809463455852)
(2, 4, 0.6970214171667677)
(3, 0, 0.5130559674850138)
(3, 1, 0.43196437834607615)
(3, 2, 0.446386627588529)
(3, 3, 0.4864293121327341)
(3, 4, 0.5411990145086231)
(4, 0, 0.5008046726069136)
(4, 1, 0.5656207899942374)
(4, 2, 0.5709920816285192)
(4, 3, 0.4394107178168875)
(4, 4, 0.5164604626143088)
```

### CODE REVIEW

**Comment by gbaraldi** [comment]
Maybe also add a test for a non empty generator?

**Comment by LilithHafner** [comment]
Most tests in test/bitset.jl construct BitSets with `BitSet(::Vector)` which covers the new code, @gbaraldi, are you referring to something else?

**Comment by gbaraldi** [comment]
Nah, I just missed it 😆 .

**Comment by KristofferC** [comment]
Is the 32bit linux failure known https://buildkite.com/julialang/julia-master/builds/15977#01835893-4af0-4ba5-8147-1cca33b597a1?

**Comment by gbaraldi** [comment]
I haven't seen that before no

**Comment by LilithHafner** [comment]
Bump and rerun CI

**Comment by LilithHafner** [comment]
Lol, closing and reopening was unnecessary.

File: base/bitset.jl [hunk 1]
```diff
@@ -16,6 +16,29 @@ mutable struct BitSet <: AbstractSet{Int}
     offset::Int
 
     BitSet() = new(sizehint!(zeros(UInt64, 0), 4), NO_OFFSET)
+    function BitSet(v::AbstractVector)
+        isempty(v) && return new(zeros(UInt64, 0), NO_OFFSET)
```
Comment (rfourquet) [inline base/bitset.jl:20]: Should this include the sizehint? Maybe
```suggestion
        isempty(v) && return BitSet()
```

**Review by rfourquet — COMMENTED** [review]
The implementation LGTM, but it will presumably be slower for `BitSet(::AbstractUnitRange{<:Integer})`, as there is a specific `union!` implementation for it. The simplest fix would probably be to add a `BitSet(::AbstractUnitRange{<:Integer})` constructor, but moving this logic to `union!` might work too (and would bring the benefit of this PR also to `union!`, although it would make things slightly more complex).

File: base/bitset.jl (hunk 1 above)
Comment (LilithHafner) [reply base/bitset.jl:20]: I don't think it should include a size hint. I think the primary reason for a size hint in the generic iterable case is because creating a BitSet with `BitSet(itr)` involves making an empty bitset and then pushing into it. We reach this case when someone ruIns `BitSet(Int[])` which may well never have new elements added to it.

**Comment by LilithHafner** [comment]
Good catch with AbstractUnitRange, and good idea about union!. The reason I made an inner constructor was to avoid double allocation, but I think it is possible to avoid that while still shifting the logic into union!.

**Comment by LilithHafner** [comment]
I moved the mechanics, though unfortunately it did add a bit more complexity.

File: base/bitset.jl [hunk 2]
```diff
@@ -123,6 +122,84 @@ end
...
+
+# assumes that every element of itr falls within the allocated bits of s
+function unsafe_union!(s::BitSet, itr)
+    for x in itr
+        unsafe_push!(s, x)
+    end
+    s
+end
+
+function BitSet(v::AbstractVector)
```
Comment (rfourquet) [inline base/bitset.jl:159]: It seems this will catch `BitSet(::AbstractUnitRange{<:Integer}`, which should still use the specialized `union!` method.

**Comment by LilithHafner** [comment]
This is too stale for me to want to pick back up. Converting to an [issue](50326).
